content/articles/geofencing-in-android-with-kotlin/index.md

```diff
@@ -0,0 +1,367 @@
+
+```kotlin
+    private fun authorizedLocation(): Boolean {
+        val formalizeForeground = (
+                PERMISSION_GRANTED == ActivityCompat.checkSelfPermission(
+                    this, Manifest.permission.ACCESS_FINE_LOCATION
+                ))
+        val formalizeBackground =
+            if (gadgetQ) {
+                PERMISSION_GRANTED == ActivityCompat.checkSelfPermission(
+                    this, Manifest.permission.ACCESS_BACKGROUND_LOCATION
+                )
+            } else {
+                true
+            }
+        return formalizeForeground && formalizeBackground
+    }
+}
+```
+
+##### Request background and fine location permissions
+This is where you request permission from the user to access their location if not granted.
+
+```kotlin
+ private fun askLocationPermission() {
+        if (authorizedLocation())
+            return
+        var grantingPermission = arrayOf(Manifest.permission.ACCESS_FINE_LOCATION)
+        val customResult = when {
+            gadgetQ -> {
+                grantingPermission += Manifest.permission.ACCESS_BACKGROUND_LOCATION
+                REQUEST_FOREGROUND_AND_BACKGROUND_PERMISSION_RESULT_CODE
+            }
+            else -> REQUEST_FOREGROUND_ONLY_PERMISSIONS_REQUEST_CODE
+        }
+        Log.d(TAG, "askLocationPermission")
+        ActivityCompat.requestPermissions(
+            this,
+            grantingPermission,
+            customResult
+        )
+    }
+```
+
+Once the user has answered the permissions request, you should process their response in `onRequestPermissionsResult()` as shown below.
+
+```kotlin
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if (requestCode == REQUEST_LOCATION_PERMISSION) {
+            if (grantResults.size > 0 && (grantResults[0] == PackageManager.PERMISSION_GRANTED))
+                startLocation()
+        }
+    }
+```
+
+#### Step 4 - Examine the gadget's location.
+Permissions granted will be worthless if the user's device location is deactivated. To verify that the device's location is enabled, add the following code.
+
+**Check the device location settings and start the Geofence**
+
+```kotlin
+private fun validateGadgetAreaInitiateGeofence(resolve: Boolean = true) {
+        val locationRequest = LocationRequest.create().apply {
+            priority = LocationRequest.PRIORITY_LOW_POWER
+        }
+        val builder = LocationSettingsRequest.Builder().addLocationRequest(locationRequest)
+
+        val client = LocationServices.getSettingsClient(this)
+        val locationResponses =
+            client.checkLocationSettings(builder.build())
```

Comment:
...
I think a lot is happening in this snippet. Add an explanation either via comments or a plain text just below the code.